### PR TITLE
Add jacobian and jacobian transpose to nmf problem

### PR DESCRIPTION
**Pull request by MohamedLaghdafHABIBOULLAH**

### CODE REVIEW

File: src/nnmf.jl [hunk 1]
```diff
@@ -63,6 +63,37 @@ function nnmf_model(m::Int = 100, n::Int = 50, k::Int = 10, T::DataType = Float6
...
 
+  function jacv!(Jv, x, v)
+    W = reshape_array(x[1:(m * k)], (m, k))
+    H = reshape_array(x[(m * k + 1):end], (k, n))
+    W_v = reshape_array(v[1:(m * k)], (m, k))
+    H_v = reshape_array(v[(m * k + 1):end], (k, n))
+    mul!(WH, W_v, H)
+    for i ∈ eachindex(WH)
+        Jv[i] = - WH[i]
+    end
```
Comment (dpo) [inline src/nnmf.jl:74]: `Jv .= .-WH`?

File: src/nnmf.jl [hunk 2]
```diff
@@ -63,6 +63,37 @@ function nnmf_model(m::Int = 100, n::Int = 50, k::Int = 10, T::DataType = Float6
     return g
   end
 
+  function jacv!(Jv, x, v)
+    W = reshape_array(x[1:(m * k)], (m, k))
+    H = reshape_array(x[(m * k + 1):end], (k, n))
+    W_v = reshape_array(v[1:(m * k)], (m, k))
+    H_v = reshape_array(v[(m * k + 1):end], (k, n))
```
Comment (dpo) [inline src/nnmf.jl:70]: Could we preallocate those outside of `jacv!()`?

File: src/nnmf.jl [hunk 3]
```diff
@@ -63,6 +63,37 @@ function nnmf_model(m::Int = 100, n::Int = 50, k::Int = 10, T::DataType = Float6
...
+    mul!(WH, W_v, H)
+    for i ∈ eachindex(WH)
+        Jv[i] = - WH[i]
+    end
+    mul!(WH, W, H_v)
+    for i ∈ eachindex(WH)
+        Jv[i] += - WH[i]
+    end
```
Comment (dpo) [inline src/nnmf.jl:78]: `Jv .-= WH`

File: src/nnmf.jl [hunk 4]
```diff
@@ -63,6 +63,37 @@ function nnmf_model(m::Int = 100, n::Int = 50, k::Int = 10, T::DataType = Float6
...
+    H_T = reshape_array(x[(m * k + 1):end], (k, n))'
+    X_v = reshape_array(w[1:(m * n)], (m, n))
+    mul!(gw, X_v, H_T)
+    mul!(gh, W_T, X_v)
+    for i ∈ eachindex(gw)
+        Jtv[i] = - gw[i]
+    end
+    for i ∈ eachindex(gh)
+        Jtv[i + m * k] = - gh[i]
+    end
```
Comment (dpo) [inline src/nnmf.jl:93]: Same comments here.

File: src/nnmf.jl [hunk 5]
```diff
@@ -71,6 +102,16 @@ function nnmf_model(m::Int = 100, n::Int = 50, k::Int = 10, T::DataType = Float6
     lvar = zeros(eltype(A), k * (m + n)),
     uvar = fill!(zeros(eltype(A), k * (m + n)), Inf),
   ),
+  FirstOrderNLSModel(
+    resid!, 
+    jacv!, 
+    jactv!, 
+    m * n, 
+    3 * rand(eltype(A), k * (m + n)),
```
Comment (dpo) [inline src/nnmf.jl:110]: Use the same `x0` in both models.

File: test/runtests.jl [hunk 6]
```diff
@@ -139,10 +139,16 @@ end
...
   @test typeof(model) <: FirstOrderModel
   @test typeof(sol) == typeof(model.meta.x0)
+  @test typeof(nls_model) <: FirstOrderNLSModel
+  @test model.meta.nvar == nls_model.meta.nvar
+  @test nls_model.nls_meta.nequ == m * n
   @test all(model.meta.lvar .== 0)
   @test all(model.meta.uvar .== Inf)
+  @test all(nls_model.meta.lvar .== 0)
+  @test all(nls_model.meta.uvar .== Inf)
+  test_objectives(model, nls_model)
```
Comment (dpo) [inline test/runtests.jl:153]: Could you also test that the jprod and jtprod match the gradient?

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/nnmf.jl (hunk 2 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply src/nnmf.jl:70]: I don't believe so, as we have to extract them from both variables x and v (see grad! implementation)

File: src/nnmf.jl (hunk 1 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply src/nnmf.jl:74]: Ok

File: test/runtests.jl (hunk 6 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply test/runtests.jl:153]: test_objectives already tests that jtprod times the residual = gradient

File: src/nnmf.jl (hunk 4 above)
Comment (MohamedLaghdafHABIBOULLAH) [reply src/nnmf.jl:93]: how exactly ?

File: src/nnmf.jl (hunk 2 above)
Comment (dpo) [reply src/nnmf.jl:70]: Please try

File: src/nnmf.jl (hunk 1 above)
Comment (dpo) [reply src/nnmf.jl:74]: @MohamedLaghdafHABIBOULLAH Please update this PR.

**Comment by dpo** [comment]
Thank you!

**Pull request merged**